go/cmd/dolt/commands/cherry-pick.go

```diff
@@ -109,173 +113,149 @@ func (cmd CherryPickCmd) Exec(ctx context.Context, commandStr string, args []str
+		return fmt.Errorf("error: failed to get dolt status: %w", err)
 	}
-	workingRoot, err := dEnv.WorkingRoot(ctx)
-	if err != nil {
-		return errhand.VerboseErrorFromError(err)
+	if hasStagedChanges {
+		return fmt.Errorf("Please commit your staged changes before using cherry-pick.")
 	}
-	stagedRoot, err := dEnv.StagedRoot(ctx)
-	if err != nil {
-		return errhand.VerboseErrorFromError(err)
-	}
-	headHash, err := headRoot.HashOf()
-	if err != nil {
-		return errhand.VerboseErrorFromError(err)
-	}
-	workingHash, err := workingRoot.HashOf()
-	if err != nil {
-		return errhand.VerboseErrorFromError(err)
-	}
-	stagedHash, err := stagedRoot.HashOf()
-	if err != nil {
-		return errhand.VerboseErrorFromError(err)
-	}
-
-	if !headHash.Equal(stagedHash) {
-		return errhand.BuildDError("Please commit your staged changes before using cherry-pick.").Build()
-	}
-
-	if !headHash.Equal(workingHash) {
-		return errhand.BuildDError("error: your local changes would be overwritten by cherry-pick.\nhint: commit your changes (dolt commit -am \"<message>\") or reset them (dolt reset --hard) to proceed.").Build()
+	if hasUnstagedChanges {
+		return fmt.Errorf("error: your local changes would be overwritten by cherry-pick.\nhint: commit your changes (dolt commit -am \"<message>\") or reset them (dolt reset --hard) to proceed.")
 	}
 
-	mergeResult, commitMsg, err := mergeCherryPickedCommit(ctx, dEnv, workingRoot, cherryStr)
+	_, err = GetRowsForSql(queryist, sqlCtx, "set @@dolt_allow_commit_conflicts = 1")
 	if err != nil {
-		return errhand.VerboseErrorFromError(err)
+		return fmt.Errorf("error: failed to set @@dolt_allow_commit_conflicts: %w", err)
 	}
-	newWorkingRoot := mergeResult.Root
-	mapOfMergeStats := mergeResult.Stats
 
-	workingHash, err = newWorkingRoot.HashOf()
+	_, err = GetRowsForSql(queryist, sqlCtx, "set @@dolt_force_transaction_commit = 1")
 	if err != nil {
-		return errhand.VerboseErrorFromError(err)
+		return fmt.Errorf("error: failed to set @@dolt_force_transaction_commit: %w", err)
 	}
 
-	if headHash.Equal(workingHash) {
-		cli.Println("No changes were made.")
-		return nil
+	q, err := dbr.InterpolateForDialect("call dolt_cherry_pick(?)", []interface{}{cherryStr}, dialect.MySQL)
+	if err != nil {
+		return fmt.Errorf("error: failed to interpolate query: %w", err)
 	}
-
-	err = dEnv.UpdateWorkingRoot(ctx, newWorkingRoot)
+	rows, err := GetRowsForSql(queryist, sqlCtx, q)
 	if err != nil {
-		return errhand.VerboseErrorFromError(err)
+		errorText := err.Error()
+		switch {
+		case "no changes were made, nothing to commit" == errorText:
+			cli.Println("No changes were made.")
+			return nil
+		case
+			strings.HasPrefix(errorText, "Merge conflict detected, transaction rolled back."),
+			strings.HasPrefix(errorText, "Committing this transaction resulted in a working set with constraint violations"):
+			return ErrCherryPickConflictsOrViolations.New()
+		case "cherry-picking a merge commit is not supported" == errorText:
```

Comment:
Yikes. Very frail. Options to make this less frail? I wonder if there is a better approach that we can fast follow with here.

Reply:
I took another look at `strings.HasPrefix(errorText, "Merge conflict detected,` and `strings.HasPrefix(errorText, "Committing`, and these changes are not required, we handle these cases already, removing these lines.

The remaining frailty has to do with our expectation of specific error messages:
1. procedure `cherry_pick` returns `no changes were made, nothing to commit` when there are no changes, but our tests are expecting "No changes were made." If we're OK with the CLI output text changing slightly, we can remove this check and just return the produces error
2. `cherry-picking a merge commit is not supported` is similar: the tests expect a `.` at the end of the message. If we're OK with removing that period, this check can also be removed. (We can also error out earlier, if we check the commit hash and identify that it belongs to a merge commit, but that's an extra query.)

If we're OK with a minor error text change, this entire block can be removed.

Reply:
Let's remove the period and start with an uppercase "N"